controllers/provisioners/eksfargate/eksfargate.go

```diff
@@ -0,0 +1,290 @@
+
+	return ctx
+}
+
+func (ctx *InstanceGroupContext) processParameters() {
+	var (
+		instanceGroup = ctx.GetInstanceGroup()
+		spec          = instanceGroup.GetEKSFargateSpec()
+		params        = make(map[string]interface{})
+	)
+
+	params["ClusterName"] = spec.GetClusterName()
+	params["ProfileName"] = spec.GetProfileName()
+	params["Selectors"] = CreateFargateSelectors(spec.GetSelectors())
+	params["Subnets"] = CreateFargateSubnets(spec.GetSubnets())
+	params["Tags"] = CreateFargateTags(spec.GetTags())
+	ctx.AwsWorker.Parameters = params
+}
+
+func CreateFargateSubnets(subnets []string) []*string {
+	stringReferences := []*string{}
+	for _, s := range subnets {
+		temp := new(string)
+		*temp = s
+		stringReferences = append(stringReferences, temp)
+	}
+	return stringReferences
+}
+func CreateFargateTags(tagArray []map[string]string) map[string]*string {
+	tags := make(map[string]*string)
+	for _, t := range tagArray {
+		for k, v := range t {
+			va := new(string)
+			*va = v
+			tags[k] = va
+		}
+	}
+	return tags
+}
+
+// Convienence function to convert from json to API.
+func CreateFargateSelectors(selectors []v1alpha1.EKSFargateSelectors) []*eks.FargateProfileSelector {
+	var eksSelectors []*eks.FargateProfileSelector
+	for _, selector := range selectors {
+		m := make(map[string]*string)
+		for k, v := range selector.Labels {
+			vv := new(string)
+			*vv = v
+			m[k] = vv
+		}
+		eksSelectors = append(eksSelectors, &eks.FargateProfileSelector{Namespace: &selector.Namespace, Labels: m})
+	}
+	return eksSelectors
+}
+func IsRetryable(instanceGroup *v1alpha1.InstanceGroup) bool {
+	for _, state := range NonRetryableStates {
+		if state == instanceGroup.GetState() {
+			return false
+		}
+	}
+	return true
+}
+func (ctx *InstanceGroupContext) Create() error {
+	var arn string
+	instanceGroup := ctx.GetInstanceGroup()
+	spec := instanceGroup.GetEKSFargateSpec()
+	if instanceGroup.GetEKSFargateSpec().GetPodExecutionRoleArn() == "" {
+		created, err := ctx.AwsWorker.CreateDefaultRole()
+		if created || err != nil {
+			ctx.Log.Info("Creating default role")
+			return err
+		}
+		role, err := ctx.AwsWorker.GetDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get default role", "error", err)
+			return err
+		}
+		arn = *role.Arn
+
+		err = ctx.AwsWorker.AttachDefaultPolicyToDefaultRole()
+		if err != nil {
+			ctx.Log.Info("Failed to get attach policy to role", "error", err)
+			return err
+		}
+		ctx.Log.Info("Attached default policy to role")
+
+	} else {
+		arn = spec.GetPodExecutionRoleArn()
+	}
+
+	ctx.Log.Info("Creating a profile.", "arn", arn)
+	tryAgain, err := ctx.AwsWorker.CreateProfile(arn)
+	if tryAgain {
+		ctx.Log.Info("Resource inuse on Create.")
+		return nil
+	}
+	if err != nil {
+		ctx.Log.Info("Creation of the fargate profile failed", "cluster",
+			spec.GetClusterName(),
+			"profile",
+			spec.GetProfileName(),
+			"error", err)
+	} else {
+		instanceGroup.SetState(v1alpha1.ReconcileModifying)
```

Comment:
I think you can skip setting this - this is the state at which it enters `Create()` and if you don't modify it it will stay in that state

Reply:
Create() is entered when the state == v1alpha.ReconcileInitCreate, Right?
This logic sets the state to Modifying (from Create()) after the profile creation is successfully started.  If the profile creation fails, it stays in the InitCreate state.

Reply:
All the creation of resources (role, policy attach and profile creation) happens in the InitCreate state, only after the last resource (in this case the profile) is created, does the state transition to ReconcileModifying...

Reply:
I see, in this case, you should just have this SetState before you `return nil` at the end, given you make the above changes with `tryAgain`.
If all the calls inside Create() are idempotent, it's OK if it requeues while it's InitCreate, it will just retry the create.